## 4.4 Data Wrangling & Subsetting

### This script contains the following points:
    1. Importing libraries
    2. Importing data
    3. Data wrangling
    4. Create data dictionary
    5. Subsetting
    6. Task - Answers to questions
    7. Exporting dataframes

### 1. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 2. Importing data

In [2]:
path = r'/Users/jodikim/I0_2022_Instacart_Analysis'

In [3]:
# Importing orders.csv
df_ords = pd.read_csv(os.path.join(path, '02_Data', 'Original_Data', 'orders.csv'), index_col = False)

In [4]:
# Importing products.csv
df_prods = pd.read_csv(os.path.join(path, '02_Data', 'Original_Data', 'products.csv'), index_col = False)

In [5]:
# Importing departments.csv
df_dep = pd.read_csv(os.path.join(path, '02_Data', 'Original_Data', 'departments.csv'), index_col = False)

### 3. Data Wrangling

In [6]:
# Dropping eval_set column from orders.csv
df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [7]:
df_ords_2.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [8]:
# Renaming column 'dow' to 'orders_day_of_week' from orders.csv
df_ords_2.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [9]:
df_ords_2.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [10]:
# Changing data type 'order_id' to a 'str' from orders.csv
df_ords_2['order_id'] = df_ords_2['order_id'].astype('str')

In [11]:
df_ords_2['order_id'].dtype

dtype('O')

In [12]:
# Changing data type 'user_id' to a 'str' from orders.csv
df_ords_2['user_id'] = df_ords_2['user_id'].astype('str')

In [13]:
df_ords_2['user_id'].dtype

dtype('O')

In [14]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [15]:
# Transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [16]:
# Creating a new dataframe from the transposed version
df_dep_t = df_dep.T

In [17]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [18]:
# Adding an index
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [19]:
# Creating a new header. Take the first row of df_dep_t for the header
new_header = df_dep_t.iloc[0]

In [20]:
new_header

0    department
Name: department_id, dtype: object

In [21]:
# Take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

In [22]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [23]:
# Set the header row as the df header
df_dep_t_new.columns = new_header

In [24]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

### 4. Data Dictionary

In [25]:
# Turning dataframe into a dictionary
data_dict = df_dep_t_new.to_dict('index')

In [26]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [27]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [28]:
print(data_dict.get('19'))

{'department': 'snacks'}


### 5. Subsetting

In [29]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [30]:
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [31]:
# Creating a new dataframe called df_snacks from df_prods
df_snacks =  df_prods[df_prods['department_id']==19]

In [32]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

### 6. Task - Answers to questions

In [33]:
df_ords_2.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [34]:
# Answer to question 2: Changing data type 'order_number' to a 'str' from orders.csv
df_ords_2['order_number'] = df_ords_2['order_number'].astype('str')

In [35]:
df_ords_2['order_number'].dtype

dtype('O')

#### Answer to question 3: Already changed column 'dow' to 'orders_day_of_week' on line 8. All other columns are intuitive¶

In [36]:
# Finding frequency for busiest hour for placing orders on df_ords
df_ords_2.head()

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1            1                   2                  8   
1  2398795       1            2                   3                  7   
2   473747       1            3                   3                 12   
3  2254736       1            4                   4                  7   
4   431534       1            5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [37]:
df_ords_2['order_hour_of_day'].describe()

count    3.421083e+06
mean     1.345202e+01
std      4.226088e+00
min      0.000000e+00
25%      1.000000e+01
50%      1.300000e+01
75%      1.600000e+01
max      2.300000e+01
Name: order_hour_of_day, dtype: float64

In [38]:
# Changing variable from float to integer
df_ords_2['order_hour_of_day'] = df_ords['order_hour_of_day'].astype('int')

In [39]:
df_ords_2['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

#### Answer to question 4: 10 is the busiest hour of the day with 288418 orders¶

In [40]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

#### Answer to question 5: Value 4 is produce

In [41]:
# Creating a subset for department '14' called 'df_breakfast'
df_breakfast =  df_prods[df_prods['department_id']==14]

In [42]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

In [43]:
# Creating a subset for department '5' called 'alcohol'
df_alcohol =  df_prods[df_prods['department_id']==5]

In [44]:
df_alcohol.head()

product_id                    product_name  aisle_id  department_id  \
51           52             Mirabelle Brut Rose       134              5   
118         119          Chardonnay Paso Robles        62              5   
149         150                       Brut Rosé       134              5   
233         234               Tennessee Whiskey       124              5   
248         249  Pinot Grigio, California, 2010        62              5   

     prices  
51     14.4  
118     5.5  
149    12.9  
233     3.1  
248     2.7

In [45]:
# Creating a subset for department '20' called 'deli'
df_deli =  df_prods[df_prods['department_id']==20]

In [46]:
df_deli.head()

product_id                                   product_name  aisle_id  \
48           49  Vegetarian Grain Meat Sausages Italian - 4 CT        14   
61           62        Premium Deli Oven Roasted Turkey Breast        96   
73           74                Artisan Chick'n & Apple Sausage        14   
84           85                           Soppressata Piccante        96   
108         109                         Grape Leaf Hummus Wrap        13   

     department_id  prices  
48              20    10.1  
61              20    14.6  
73              20     3.0  
84              20     8.9  
108             20     7.1

In [47]:
# Creating a subset for department '7' called 'bevs'
df_bevs =  df_prods[df_prods['department_id']==7]

In [48]:
df_bevs.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
19          20  Pomegranate Cranberry & Aloe Vera Enrich Drink        98   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
19              7     6.0

In [49]:
# Creating a subset for department '12' called 'meat_seafood'
df_meat_seafood =  df_prods[df_prods['department_id']==12]

In [50]:
df_meat_seafood.head()

product_id                                       product_name  aisle_id  \
16          17                                  Rendered Duck Fat        35   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   
83          84                                         Lamb Shank         7   

    department_id  prices  
16             12    17.1  
22             12     8.2  
34             12    15.1  
39             12    22.5  
83             12    24.3

In [51]:
# Creating a subset for departments 5, 20, 7, 12 called 'df_dinner_parties'
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [52]:
df_dinner_parties.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

In [53]:
# Answer to question 8: Calculating the rows of df_dinner_parties
df_dinner_parties.shape

(7650, 5)

In [55]:
# Extracting information about user_id 1
df_ords_user_1 = df_ords_2[df_ords_2['user_id']=='1']

In [57]:
df_ords_user_1

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1            1                   2                  8   
1   2398795       1            2                   3                  7   
2    473747       1            3                   3                 12   
3   2254736       1            4                   4                  7   
4    431534       1            5                   4                 15   
5   3367565       1            6                   2                  7   
6    550135       1            7                   1                  9   
7   3108588       1            8                   1                 14   
8   2295261       1            9                   1                 16   
9   2550362       1           10                   4                  8   
10  1187899       1           11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [58]:
df_ords_user_1.describe()

orders_day_of_week  order_hour_of_day  days_since_prior_order
count           11.000000          11.000000               10.000000
mean             2.636364          10.090909               19.000000
std              1.286291           3.477198                9.030811
min              1.000000           7.000000                0.000000
25%              1.500000           7.500000               14.250000
50%              3.000000           8.000000               19.500000
75%              4.000000          13.000000               26.250000
max              4.000000          16.000000               30.000000

### 7. Exporting dataframes

In [59]:
# Exporting df_ords_2 dataframes 
df_ords_2.to_csv(os.path.join(path, '02_Data','Prepared_Data', 'orders_wrangled.csv'))

In [60]:
# Exporting df_dep_t_new dataframes 
df_dep_t_new.to_csv(os.path.join(path, '02_Data','Prepared_Data', 'departments_wrangled.csv'))